packages/angular/build/src/utils/server-rendering/manifest.ts

```diff
@@ -138,15 +146,89 @@ export function generateAngularServerAppManifest(
     }
   }
 
+  // When routes have been extracted, mappings are no longer needed, as preloads will be included in the metadata.
+  // When shouldOptimizeChunks is enabled the metadata is no longer correct and thus we cannot generate the mappings.
+  const serverToBrowserMappings =
+    routes?.length || shouldOptimizeChunks
+      ? undefined
+      : generateLazyLoadedFilesMappings(metafile, initialFiles, publicPath);
+
   const manifestContent = `
 export default {
   bootstrap: () => import('./main.server.mjs').then(m => m.default),
   inlineCriticalCss: ${inlineCriticalCss},
+  locale: ${JSON.stringify(locale, undefined, 2)},
+  serverToBrowserMappings: ${JSON.stringify(serverToBrowserMappings, undefined, 2)},
   routes: ${JSON.stringify(routes, undefined, 2)},
   assets: new Map([\n${serverAssetsContent.join(', \n')}\n]),
-  locale: ${locale !== undefined ? `'${locale}'` : undefined},
 };
 `;
 
   return { manifestContent, serverAssetsChunks };
 }
+
+/**
+ * Generates a mapping of lazy-loaded files from a given metafile.
+ *
+ * This function processes the outputs of a metafile to create a mapping
+ * between MJS files (server bundles) and their corresponding JS files
+ * that should be lazily loaded. It filters out files that do not have
+ * an entry point, do not export any modules, or are not of the
+ * appropriate file extensions (.js or .mjs).
+ *
+ * @param metafile - An object containing metadata about the output files,
+ * including entry points, exports, and imports.
+ * @param initialFiles - A set of initial file names that are considered
+ * already loaded and should be excluded from the mapping.
+ * @param publicPath - The configured public path.
+ *
+ * @returns A record where the keys are MJS file names (server bundles) and
+ * the values are arrays of corresponding JS file names (browser bundles).
+ */
+function generateLazyLoadedFilesMappings(
+  metafile: Metafile,
+  initialFiles: Set<string>,
+  publicPath = '',
+): Record<string, readonly string[]> {
+  const entryPointToBundles = new Map<
+    string,
+    { js: string[] | undefined; mjs: string | undefined }
+  >();
+
+  for (const [fileName, { entryPoint, exports, imports }] of Object.entries(metafile.outputs)) {
+    const extension = extname(fileName);
+
+    // Skip files that don't have an entryPoint, no exports, or are not .js or .mjs
+    if (!entryPoint || exports?.length < 1 || (extension !== '.js' && extension !== '.mjs')) {
+      continue;
+    }
+
+    const data = entryPointToBundles.get(entryPoint) ?? { js: undefined, mjs: undefined };
+    if (extension === '.js') {
```

Comment:
The mjs are the server files whilst the js are the browser files.

Reply:
Is that a safe assumption to make? If we ever change file extensions, that would break this logic. Is there a different mechanism which could separate these files?

Reply:
We could generate separate metadata for server and browser builds. There is a feature request for that.